File: src/main/java/core/gui/ExceptionDialog.java
package core.gui;

import core.gui.theme.HOIconName;
import core.gui.theme.ThemeManager;
import core.util.ExceptionUtils;
import core.util.HOLogger;

import java.awt.BorderLayout;
import java.awt.Dimension;
import java.awt.Font;
import java.awt.GridBagConstraints;
import java.awt.GridBagLayout;
import java.awt.Insets;
import java.awt.event.ActionEvent;
import java.awt.event.ActionListener;

import javax.swing.ImageIcon;
import javax.swing.JButton;
import javax.swing.JDialog;
import javax.swing.JLabel;
import javax.swing.JPanel;
import javax.swing.JScrollPane;
import javax.swing.JTextArea;

/**
 * 
 * @author kruescho
 */
public class ExceptionDialog extends JDialog {

	private static final long serialVersionUID = 767279628764672228L;
	private JPanel detailsPanel;
	private JTextArea textArea;
	private JButton detailsButton;
	private Dimension savedDetailsSize;
	private ImageIcon showDetailsImage;
	private ImageIcon hideDetailsImage;
	private Throwable throwable;

	public ExceptionDialog(String msg, Throwable t) {
		setModal(true);
		this.throwable = t;
		setTitle(msg);
		setDefaultCloseOperation(JDialog.DISPOSE_ON_CLOSE);
		initComponents();
	}

	private void initComponents() {
		try {
			this.showDetailsImage = ThemeManager.getIcon(HOIconName.CONTROL_DOUBLE_270);
			this.hideDetailsImage = ThemeManager.getIcon(HOIconName.CONTROL_DOUBLE_090);
			setIconImage(ThemeManager.getIcon(HOIconName.EXCLAMATION_RED).getImage());
		} catch (Exception ex) {
			HOLogger.instance().log(getClass(), ex);
		}
		setLayout(new BorderLayout());

		JPanel topPanel = new JPanel();
		topPanel.setLayout(new GridBagLayout());

		JLabel label = new JLabel("An error occurred.");
		label.setFont(label.getFont().deriveFont(label.getFont().getStyle() ^ Font.BOLD));
		try {
			label.setIcon(ThemeManager.getIcon(HOIconName.EXCLAMATION_RED));
		} catch (Exception ex) {
			HOLogger.instance().log(getClass(), ex);
		}
		GridBagConstraints gbc = new GridBagConstraints();
		gbc.insets = new Insets(8, 8, 4, 4);
		gbc.anchor = GridBagConstraints.WEST;
		topPanel.add(label, gbc);

		JLabel errorLabel = new JLabel(getMessage(this.throwable));
		gbc = new GridBagConstraints();
		gbc.gridy = 1;
		gbc.insets = new Insets(4, 8, 8, 4);
		gbc.anchor = GridBagConstraints.WEST;
		topPanel.add(errorLabel, gbc);

		this.detailsButton = new JButton("details", this.showDetailsImage);
		gbc = new GridBagConstraints();
		gbc.insets = new Insets(4, 4, 8, 8);
		gbc.gridy = 1;
		gbc.gridx = 1;
		gbc.anchor = GridBagConstraints.WEST;
		gbc.weightx = 1.0;
		topPanel.add(this.detailsButton, gbc);
		this.detailsButton.addActionListener(new ActionListener() {

			@Override
			public void actionPerformed(ActionEvent e) {
				switchDetails();
			}
		});

		add(topPanel, BorderLayout.NORTH);
		pack();
	}

	private String getMessage(Throwable throwable) {
		String message = throwable.getMessage();
		if (message == null) {
			message = throwable.getClass().getName();
		}
		return message;
	}

	private void createDetails() {
		this.detailsPanel = new JPanel(new BorderLayout());
		this.textArea = new JTextArea();
		textArea.setEditable(false);
		textArea.setText(ExceptionUtils.getStackTrace(this.throwable));
		JScrollPane scrollPane = new JScrollPane(textArea);
		this.detailsPanel.add(scrollPane, BorderLayout.CENTER);
		this.detailsPanel.setPreferredSize(new Dimension(650, 300));
		this.savedDetailsSize = new Dimension(this.detailsPanel.getPreferredSize());
		add(this.detailsPanel, BorderLayout.CENTER);
	}

	private void switchDetails() {
		if (this.detailsPanel != null && this.detailsPanel.isVisible()) {
			this.savedDetailsSize = new Dimension(this.detailsPanel.getSize());
			this.detailsButton.setIcon(this.showDetailsImage);
			this.detailsPanel.setVisible(false);
		} else {
			if (this.detailsPanel == null) {
				createDetails();
			} else {
				this.detailsPanel.setPreferredSize(this.savedDetailsSize);
			}
			this.detailsButton.setIcon(this.hideDetailsImage);
			this.detailsPanel.setVisible(true);
			this.textArea.setCaretPosition(0);
		}
		pack();
	}
}

File: src/main/java/core/gui/ExceptionDialog.kt
package core.gui

import core.gui.theme.HOIconName
import core.gui.theme.ImageUtilities
import core.gui.theme.ThemeManager
import core.util.ExceptionUtils
import core.util.HOLogger
import java.awt.*
import javax.swing.*

/**
 *
 * @author kruescho
 */
class ExceptionDialog(msg: String?, t: Throwable) : JDialog() {
    private var detailsPanel: JPanel? = null
    private var textArea: JTextArea? = null
    private var detailsButton: JButton? = null
    private var savedDetailsSize: Dimension? = null
    private var showDetailsImage: Icon? = null
    private var hideDetailsImage: Icon? = null
    private val throwable: Throwable

    init {
        setModal(true)
        throwable = t
        setTitle(msg)
        setDefaultCloseOperation(DISPOSE_ON_CLOSE)
        initComponents()
    }

    private fun initComponents() {
        try {
            showDetailsImage = ThemeManager.getIcon(HOIconName.CONTROL_DOUBLE_270)
            hideDetailsImage = ThemeManager.getIcon(HOIconName.CONTROL_DOUBLE_090)
            setIconImage(ImageUtilities.iconToImage(ThemeManager.getIcon(HOIconName.EXCLAMATION_RED)))
        } catch (ex: Exception) {
            HOLogger.instance().log(javaClass, ex)
        }
        layout = BorderLayout()
        val topPanel = JPanel()
        topPanel.setLayout(GridBagLayout())
        val label = JLabel("An error occurred.")
        label.setFont(label.font.deriveFont(label.font.style xor Font.BOLD))
        try {
            label.setIcon(ThemeManager.getIcon(HOIconName.EXCLAMATION_RED))
        } catch (ex: Exception) {
            HOLogger.instance().log(javaClass, ex)
        }
        var gbc = GridBagConstraints()
        gbc.insets = Insets(8, 8, 4, 4)
        gbc.anchor = GridBagConstraints.WEST
        topPanel.add(label, gbc)
        val errorLabel = JLabel(getMessage(throwable))
        gbc = GridBagConstraints()
        gbc.gridy = 1
        gbc.insets = Insets(4, 8, 8, 4)
        gbc.anchor = GridBagConstraints.WEST
        topPanel.add(errorLabel, gbc)
        detailsButton = JButton("details", showDetailsImage)
        gbc = GridBagConstraints()
        gbc.insets = Insets(4, 4, 8, 8)
        gbc.gridy = 1
        gbc.gridx = 1
        gbc.anchor = GridBagConstraints.WEST
        gbc.weightx = 1.0
        topPanel.add(detailsButton, gbc)
        detailsButton!!.addActionListener { switchDetails() }
        add(topPanel, BorderLayout.NORTH)
        pack()
    }

    private fun getMessage(throwable: Throwable): String? {
        var message = throwable.message
        if (message == null) {
            message = throwable.javaClass.getName()
        }
        return message
    }

    private fun createDetails() {
        detailsPanel = JPanel(BorderLayout())
        textArea = JTextArea()
        textArea!!.isEditable = false
        textArea!!.text = ExceptionUtils.getStackTrace(throwable)
        val scrollPane = JScrollPane(textArea)
        detailsPanel!!.add(scrollPane, BorderLayout.CENTER)
        detailsPanel!!.preferredSize = Dimension(650, 300)
        savedDetailsSize = Dimension(detailsPanel!!.getPreferredSize())
        add(detailsPanel, BorderLayout.CENTER)
    }

    private fun switchDetails() {
        if (detailsPanel != null && detailsPanel!!.isVisible) {
            savedDetailsSize = Dimension(detailsPanel!!.size)
            detailsButton!!.setIcon(showDetailsImage)
            detailsPanel!!.isVisible = false
        } else {
            if (detailsPanel == null) {
                createDetails()
            } else {
                detailsPanel!!.preferredSize = savedDetailsSize
            }
            detailsButton!!.setIcon(hideDetailsImage)
            detailsPanel!!.isVisible = true
            textArea!!.setCaretPosition(0)
        }
        pack()
    }
}
